operator/src/main/java/org/bf2/operator/managers/CapacityManager.java

```diff
@@ -0,0 +1,330 @@
+            }
+        }
+    }
+
+    static String getManagedKafkaKey(ManagedKafka managedKafka) {
+        return MANAGED_KAFKA_PREFIX + managedKafka.getId();
+    }
+
+    private boolean releaseResources(Map<String, String> resourceData, String entryKey) {
+        String previous = resourceData.remove(entryKey);
+        if (previous != null) {
+            Resources resources = Serialization.unmarshal(previous, Resources.class);
+            resourceData.computeIfPresent(resources.profile,
+                    (k, v) -> String.valueOf(Integer.parseInt(v) - resources.units));
+            return true;
+        }
+        return false;
+    }
+
+    public synchronized Optional<OperandReadiness> claimResources(ManagedKafka managedKafka, String profile,
+            ManagedKafkaAgent agent) {
+        Integer max = getProfileMaxUnits(agent.getSpec().getCapacity(), profile).orElse(null);
+        if (max == null) {
+            return Optional.empty();
+        }
+
+        // ensure we have the latest
+        ConfigMap resourceConfigMap = client.configMaps().withName(FLEETSHARD_RESOURCES).get();
+
+        Resources resources = createResources(managedKafka, profile);
+
+        int total = Integer.parseInt(resourceConfigMap.getData().getOrDefault(profile, "0"));
+
+        int newTotal = resources.units + total;
+        if (newTotal > max) {
+            log.warnf("Rejecting the deployment of %s/%s as the cluster is full",
+                    managedKafka.getMetadata().getNamespace(), managedKafka.getMetadata().getName());
+            return Optional
+                    .of(new OperandReadiness(Status.False, Reason.Rejected, "Cluster has insufficient resources"));
+        }
+
+        // update the configmap in a locked manner - this may fail, but we'll retry
+        resourceConfigMap.getData().put(profile, String.valueOf(newTotal));
+        resourceConfigMap.getData().put(getManagedKafkaKey(managedKafka), Serialization.asJson(resources));
+        client.configMaps()
+                .withName(FLEETSHARD_RESOURCES)
+                .lockResourceVersion(resourceConfigMap.getMetadata().getResourceVersion())
+                .replace(resourceConfigMap);
+        return Optional.empty();
+    }
+
+    /**
+     * For now we are operating under the assumption that nothing can change about the resources used by a managed kafka
+     *
+     * In the future, this will need to be relaxed
+     */
+    Optional<OperandReadiness> validateResources(ManagedKafka managedKafka, String profile, String claimedResources) {
+        Resources resources = Serialization.unmarshal(claimedResources, Resources.class);
+
+        String message = null;
+
+        if (managedKafka.isReserveDeployment()) {
+            message = String.format("The deployment type for %s/%s was flipped to reserved",
+                    managedKafka.getMetadata().getNamespace(), managedKafka.getMetadata().getName());
+        }
```

Comment:
This message is shown even when the MK instance in argument was a reserved instance.

```
The deployment type for reserved-kafka-standard-1/reserved-kafka-standard-1 was flipped to reserved
```

Reply:
Yes, that did not translate well from some earlier logic merge.  For the purposes of validation we don't actually care if something flips from a real instance to reserved - it will be the same amount either way.  We need either a higher level enforcement, or logic in KafkaCluster to handle this case.

Reply:
After trying to move this handling it became clear that it was quicker to just leave this check but pair it with another one that prevents the claim for reserved managed kafkas.

This does leave us with a corner case situation which is if you have a reserved deployment that you change to a real one - this will currently leave the reserved Deployments in place if they are already created.  We'll address this in that other JIRA about corner cases.

Reply:
Sounds good. +1, I thought adding handling of reserved instance would suffice to take care of this as well.